File: Projector.py
import torch
from torch.autograd import Variable
import torch.nn.functional as F
import torch.utils.data as Data
import torch_optimizer as optim
import random
from torch.utils.tensorboard import SummaryWriter
import numpy as np

import time

start_time = time.time()

projectorIn = []
projectorOut = []
latent = []

with open("/home/pau1o-hs/Documents/Database/XData.txt") as f:
    for line in f:
        inner_list = [elt.strip() for elt in line.split(' ')]

        if inner_list == ['']:
            continue

        converted =  np.asarray(inner_list, dtype=np.float64, order='C')
        projectorIn.append(converted)

with open("/home/pau1o-hs/Documents/Database/LatentVariables.txt") as f:
    for line in f:
        inner_list = [elt.strip() for elt in line.split(' ')]
        
        if inner_list == ['']:
            continue
        
        converted =  np.asarray(inner_list, dtype=np.float64, order='C')
        latent.append(converted)

device = torch.device("cuda")

projectorOut = np.append(projectorIn, latent, axis=1)
x = torch.tensor(projectorIn, dtype=torch.float).to(device)
y = torch.tensor(projectorOut, dtype=torch.float).to(device)

means = x.mean(dim=1, keepdim=True)
stds = x.std(dim=1, keepdim=True)
normalized_in = (x - means) / stds

means = y.mean(dim=1, keepdim=True)
stds = y.std(dim=1, keepdim=True)
normalized_out = (y - means) / stds

torch.set_printoptions(precision=6)
np.set_printoptions(precision=7, floatmode='fixed', suppress=True)

train = Data.TensorDataset(normalized_in, normalized_out)
train_loader = Data.DataLoader(train, shuffle=True, batch_size=32)

class Net(torch.nn.Module):
    def __init__(self, n_feature, n_hidden, n_hidden2, n_hidden3, n_hidden4, n_output):
        super(Net, self).__init__()
        self.hidden = torch.nn.Linear(n_feature, n_hidden)    # hidden layer
        self.hidden2 = torch.nn.Linear(n_hidden, n_hidden2)   # hidden layer
        self.hidden3 = torch.nn.Linear(n_hidden2, n_hidden3)  # hidden layer
        self.hidden4 = torch.nn.Linear(n_hidden3, n_hidden4)  # hidden layer
        self.predict = torch.nn.Linear(n_hidden4, n_output)   # output layer

    def forward(self, x):
        x = F.relu(self.hidden(x))       # activation function for hidden layer
        x = F.relu(self.hidden2(x))      # activation function for hidden layer
        x = F.relu(self.hidden3(x))      # activation function for hidden layer
        x = F.relu(self.hidden4(x))      # activation function for hidden layer
        x = self.predict(x)              # linear output
        return x

def my_loss(noiseInput, predict, target):
    l2loss = torch.abs((noiseInput - target[:, :24].clone())**2 - (noiseInput - predict[:, :24].clone())**2)
    loss = torch.mean(torch.abs(target - predict) + torch.cat((l2loss, torch.zeros(noiseInput.size(0), 64).to(device)), 1))
    return loss

net = Net(n_feature=24, n_hidden=512, n_hidden2=512, n_hidden3=512, n_hidden4=512, n_output=24+64).to(device)     # define the network

optimizer = optim.RAdam(net.parameters(), lr=0.001)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1000, gamma=0.99)
loss_func = torch.nn.MSELoss() # this is for regression mean squared loss

writer = SummaryWriter()

# train the network
for t in range(10000):
    epochLoss = 0.0

    for batch_idx, (data, target) in enumerate(train_loader):

        data.to(device), target.to(device)

        noise = torch.randn(data.size(0), 24).to(device)*(random.uniform(0.0, 1.0)**(0.5))
        dataNoise = data + noise

        knnIndices = []
        newTargets = torch.zeros(target.size(0), target.size(1)).to(device)

        for i in range(0, len(data)):
            dist = torch.norm(normalized_in - dataNoise[i], dim=1, p=None)
            knn = dist.topk(1, largest=False)
            knnIndices.append(knn.indices)

        for i in range(0, len(data)):
            newTargets[i] = normalized_out[knnIndices[i]]

        prediction = net(dataNoise)     # input x and predict based on x

        loss = loss_func(prediction, newTargets)     # must be (1. nn output, 2. target)

        optimizer.zero_grad()   # clear gradients for next train
        loss.backward()         # backpropagation, compute gradients
        optimizer.step()        # apply gradients
        scheduler.step()

        epochLoss += loss * prediction.size(0)

    if t % 500 == 0:
        print(t, epochLoss.item())
    
    writer.add_scalar('Python Projector Loss', epochLoss, t)

# prediction = net(normalized_data[0])

# print(normalized_data[0].detach().numpy())
# print()
# print(prediction.detach().numpy())

with open('/home/pau1o-hs/Documents/NNWeights/Projector.txt', "w+") as f:
    np.savetxt(f, net.hidden.weight.cpu().detach().numpy().transpose(), delimiter="\n")    
    np.savetxt(f, net.hidden.bias.cpu().detach().numpy(), delimiter="\n")

    np.savetxt(f, net.hidden2.weight.cpu().detach().numpy().transpose(), delimiter="\n")    
    np.savetxt(f, net.hidden2.bias.cpu().detach().numpy(), delimiter="\n")

    np.savetxt(f, net.hidden3.weight.cpu().detach().numpy().transpose(), delimiter="\n")    
    np.savetxt(f, net.hidden3.bias.cpu().detach().numpy(), delimiter="\n")

    np.savetxt(f, net.hidden4.weight.cpu().detach().numpy().transpose(), delimiter="\n")    
    np.savetxt(f, net.hidden4.bias.cpu().detach().numpy(), delimiter="\n")
    
    np.savetxt(f, net.predict.weight.cpu().detach().numpy().transpose(), delimiter="\n")
    np.savetxt(f, net.predict.bias.cpu().detach().numpy(), delimiter="\n")

print("Runtime: %s minutes" % ((time.time() - start_time) / 60))

File: Projector.cpp
#include "CustomLib.hpp"

// C++ neural network model defined with value semantics
struct Net : torch::nn::Module
{
    Net(torch::Device device, int64_t n_feature, int64_t n_hidden1, int64_t  n_hidden2, int64_t  n_hidden3, int64_t  n_hidden4, int64_t n_output)
    {
        this->fc1     = register_module("fc1",     torch::nn::Linear(n_feature, n_hidden1));
        this->fc2     = register_module("fc2",     torch::nn::Linear(n_hidden1, n_hidden2));
        this->fc3     = register_module("fc3",     torch::nn::Linear(n_hidden2, n_hidden3));
        this->fc4     = register_module("fc4",     torch::nn::Linear(n_hidden3, n_hidden4));
        this->predict = register_module("predict", torch::nn::Linear(n_hidden4, n_output));

        this->fc1->to(device);
        this->fc2->to(device);
        this->fc3->to(device);
        this->fc4->to(device);
        this->predict->to(device);
    }

    torch::Tensor forward(torch::Tensor feature)
    {
        feature = torch::relu(this->fc1(feature));
        feature = torch::relu(this->fc2(feature));
        feature = torch::relu(this->fc3(feature));
        feature = torch::relu(this->fc4(feature));
        feature = this->predict(feature);
        return feature;
    }

    torch::nn::Linear fc1 = nullptr;
    torch::nn::Linear fc2 = nullptr;
    torch::nn::Linear fc3 = nullptr;
    torch::nn::Linear fc4 = nullptr;
    torch::nn::Linear predict = nullptr;
};

int main()
{
    auto start = chrono::high_resolution_clock::now();
    
	cout << "[Device: " << at::cuda::getDeviceProperties(0)->name << "]" << endl;
    cout << "[Count: " << torch::cuda::device_count() << "]" << endl << endl;

    // VARIABLES
    const int64_t n_feature  = 56;
    const int64_t n_hidden1  = 512;
    const int64_t n_hidden2  = 512;
    const int64_t n_hidden3  = 512;
    const int64_t n_hidden4  = 512;
    const int64_t n_output   = 56;
    const int64_t batch_size = 32;

    // INIT NN
    auto model = Net(device, n_feature, n_hidden1, n_hidden2, n_hidden3, n_hidden4, n_output);
    model.to(device);

    // DATASET
    vector<torch::Tensor> x, xNorm;
    LoadData("Features", x, true);

	// NORMALIZE DATA
    for (int i = 0; i < (int) x.size(); i++)
    {
        auto mean = x[i].mean(1, true);
        auto std  = x[i].std(1, true).reshape({x[i].size(0), 1});

        xNorm.push_back((x[i] - mean) / std);
    }

    auto dataset = CustomDataset(xNorm, xNorm, 1, 5).map(torch::data::transforms::Stack<>());

    // DATA LOADER
    auto data_loader = torch::data::make_data_loader<torch::data::samplers::RandomSampler>
    (std::move(dataset), torch::data::DataLoaderOptions().batch_size(batch_size).workers(4));

    // OPTIMIZER
    torch::optim::AdamW optimizer(model.parameters(), 0.001);
	torch::optim::StepLR scheduler(optimizer, 1000, 0.99);

    cout << "Training Projector...\n";
	
    for (size_t epoch = 1; epoch <= 50000; epoch++)
    {
        int batchId = 0;
        float epochLoss = 0;

        // Iterate the data loader to yield batches from the dataset.
        for (auto &batch : *data_loader)
        {
            batch.data = batch.data.to(device);
            batch.target = batch.target.to(device);
            
            // cout << batch.data.size(0) << ' ' << batch.data.size(1) << ' ' << batch.data.size(2) << endl;

            // Execute the model on the input data.
            torch::Tensor prediction = model.forward(batch.data);
            
            // Compute a loss value to judge the prediction of our model.
            torch::Tensor loss = torch::nn::MSELoss()(prediction, batch.target);

            // Reset gradients.
            optimizer.zero_grad();

            // Compute gradients of the loss w.r.t. the parameters of our model.
            loss.backward();

            // Update the parameters based on the calculated gradients.
            optimizer.step();
            scheduler.step();

            epochLoss += prediction.size(0) * loss.item<float>();
            
            // // Output the loss and checkpoint every 100 batches.
            // if (++batchId % 9 == 0 && epoch % 1000 == 0)
            // {
            //     // std::cout << "Epoch: " << epoch << " | Loss: " << loss.item<float>() << std::endl;
            //     // Serialize your model periodically as a checkpoint.
            //     // torch::save(model, "Decompressor.pt");
            // }
        }

        if (epoch % 5000 == 0)
        {
            cout << "Epoch: " << epoch << " | Loss: " << epochLoss << endl;
        }
    }

    std::cout << "Finished\n\n";

    // torch::Tensor prediction = model.forward(x[0]);
    // cout << prediction << endl;
    // cout << prediction[0].transpose(0, 0) << endl;

	// SAVE WEIGHTS
	ofstream weights("/home/pau1o-hs/Documents/NNWeights/Projector.txt");
	torch::Tensor weightData;

	// FC1
	weightData = model.fc1->weight.cpu().detach().transpose(0, 1).flatten();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	weightData = model.fc1->bias.cpu().detach();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	// FC2
	weightData = model.fc2->weight.cpu().detach().transpose(0, 1).flatten();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	weightData = model.fc2->bias.cpu().detach();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	// FC3
	weightData = model.fc3->weight.cpu().detach().transpose(0, 1).flatten();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	weightData = model.fc3->bias.cpu().detach();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	// FC4
	weightData = model.fc4->weight.cpu().detach().transpose(0, 1).flatten();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	weightData = model.fc4->bias.cpu().detach();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	// PREDICT
	weightData = model.predict->weight.cpu().detach().transpose(0, 1).flatten();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	weightData = model.predict->bias.cpu().detach();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

    // RUNTIME
    auto stop = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::seconds>(stop - start);
    cout << "Runtime: " << (float) (duration.count() / 60.0f) << " minutes" << endl;
}